### ISSUE TITLE
Multiple increment/decrement bugs on counter_cache with has_many through association

### ISSUE THREAD

**Issue by felixbuenemann** [rails/rails#23739]
There are multiple bugs affecting at least Rails 4.1.14.1, 4.2.5.1, 5.0.0.beta2 and master. Depending on the version the bugs range from temporarily double incrementing counters, fixed by reload, to counters incrementing(!) when destroying join records. 

I'll try to summarize the behavior and attach a test case that hast multiple test that fail.

|  | 4.1.14.1 | 4.2.5.1 | master |
| --- | --- | --- | --- |
| assign | 1 :+1: | 2 :-1: | 1 :+1: |
| assign + reload | 2 :-1: | 1 :+1: | 1 :+1: |
| clear by re-assign | 0 :+1: | 3 :-1: | 1 :-1: |
| clear by re-assign + reload | 1 :-1: | 1 :-1: | 1 :-1: |
| clear by destroy_all | 0 :+1: | 2 :-1: | 0 :+1: |
| clear by destroy_all + reload | 0 :+1: | 0 :+1: | 0 :+1: |
| failed | 2 | 4 | 2 |

**Here's the data model used:**

``` ruby
class Product < ActiveRecord::Base
  has_many :classifications
  has_many :taxons, through: :classifications
end

class Classification < ActiveRecord::Base
  belongs_to :product
  belongs_to :taxon, counter_cache: :products_count
end

class Taxon < ActiveRecord::Base
  has_many :classifications
  has_many :products, through: :classifications
end
```

**Test cases summarized above:**

``` ruby
class BugTest < Minitest::Test

  def taxon
    @taxon ||= Taxon.create!
  end

  def product
    @product ||= Product.create!
  end

  def test_counter_cache_initial_value
    assert_equal 0, taxon.products_count
  end

  def test_counter_cache_after_assign
    taxon.products = [product]
    assert_equal 1, taxon.products_count
  end

  def test_counter_cache_after_assign_and_reload
    taxon.products = [product]
    assert_equal 1, taxon.reload.products_count
  end

  def test_counter_cache_after_clear_by_assign
    taxon.products = [product]
    taxon.products = []
    assert_equal 0, taxon.products_count
  end

  def test_counter_cache_after_clear_by_assign_and_reload
    taxon.products = [product]
    taxon.products = []
    assert_equal 0, taxon.reload.products_count
  end

  def test_counter_cache_after_clear_by_destroy_all
    taxon.products = [product]
    taxon.products.destroy_all
    assert_equal 0, taxon.products_count
  end

  def test_counter_cache_after_clear_by_destroy_all_and_reload
    taxon.products = [product]
    taxon.products.destroy_all
    assert_equal 0, taxon.reload.products_count
  end

end
```

**Also available as self-contained test-case as a gist:**

[test_case_active_record_has_many_through_counter_cache_bugs.rb](https://gist.github.com/ce8a7c9844d579b4f624)

**Comment by felixbuenemann** [comment]
It's also worth checking the implicit counter cache behavior on 4.1/4.2, it show different wrong counts not listed above. I'll compile a table for that.

**Comment by maclover7** [comment]
Thanks for all of this information! Would you be willing to work on fixing the remaining issues on master?

**Comment by felixbuenemann** [comment]
I'll take a look at it this weekend. Ideally this would also be fixed in 4.2m it's in very bad shape. I couldn't find any combination that allowed bug free behavior on 4.2.

**Here's the table for the implicit counter cache (cache column, but no counter_cache on belongs_to):**

|                               | 4.1.14.1 | 4.2.5.1 | master |
|-------------------------------|----------|---------|--------|
| assign                        | 1 :+1:   | 1 :+1:  | 1 :+1: |
| assign + reload               | 1 :+1:   | 1 :+1:  | 1 :+1: |
| clear by re-assign            | 0 :+1:   | 2 :-1:  | 1 :-1: |
| clear by re-assign + reload   | 0 :+1:   | 1 :-1:  | 1 :-1: |
| clear by destroy_all          | 0 :+1:   | 1 :-1:  | 0 :+1: |
| clear by destroy_all + reload | 0 :+1:   | 1 :-1:  | 0 :+1: |
| failed                        | 0        | 4       | 2      |

This apparently worked fine on 4.1 while 4.2 has different errors than with explicit `counter_cache` and master is consistent with the explicit behavior.

**Here's the model definition differences for the implicit test case:**

```diff
--- explicit_counter_cache.rb	2016-02-17 19:05:05.000000000 +0100
+++ implicit_counter_cache.rb	2016-02-17 19:04:42.000000000 +0100
@@ -49,7 +49,7 @@

 class Classification < ActiveRecord::Base
   belongs_to :product
-  belongs_to :taxon, counter_cache: :products_count
+  belongs_to :taxon
 end

 class Taxon < ActiveRecord::Base
```

The implicit counter cache behavior is deprecated in rails 5.0.

**Comment by felixbuenemann** [comment]
Btw. the deprecation message for implicit counter caches mentions, that the counter_cache should be specified on both the belong_to and has_many associtions, which in reality lead to double increments on assignment:

> DEPRECATION WARNING: Automatic updating of counter caches on through associations has been deprecated, and will be removed in Rails 5. Instead, please set the appropriate `counter_cache` options on the `has_many` and `belongs_to` for your associations to classifications. (called from test_counter_cache_after_clear_by_destroy_all at test_case_active_record_has_many_through_counter_cache_bugs.rb:97)

I couldn't find a single test case that actually uses it on both the belong_to and remote side of the has_many for a has_many through relationship.

**Comment by felixbuenemann** [comment]
Issue #19550 might be related, although it uses assignment by `<<` which is not used here.

**Comment by maclover7** [comment]
Might be easier to create a gist with all relevant information, so that way we don't rack up a ton of comments here.

**Comment by maclover7** [comment]
Pooled information here --> https://gist.github.com/maclover7/e3647aac47606b26afb0

**Comment by felixbuenemann** [comment]
@maclover7 I've had a look at this today.

From what I gathered so far one problem is, that some code requires the counter_cache to be defined on the belongs_to, but that's not always true. There are tests that document that it should be possible to define the counter cache on a has_many/has_many through association. There's also the case of has_and_belongs_to_many, where it's simply not possible for the user to specify the counter_cache on the belongs_to, because it's on an anonymous class.

The easiest solution would probably to have has_many find its inverse belongs_to and automatically set the counter_cache option on it as well.

I also have a fix for the clear by re-assign bug above, which is similar. It requires looking at the inverse_of of the through_reflection, because the counter_cache is not defined on the through association:

```diff
diff --git a/activerecord/lib/active_record/associations/has_many_through_association.rb b/activerecord/lib/active_record/associations/has_many_through_association.rb
index 36fc381..ffc6c8f 100644
--- a/activerecord/lib/active_record/associations/has_many_through_association.rb
+++ b/activerecord/lib/active_record/associations/has_many_through_association.rb
@@ -162,7 +162,11 @@ def delete_records(records, method)
           end

           if through_reflection.collection? && update_through_counter?(method)
-            update_counter(-count, through_reflection)
+            if !method && through_reflection.inverse_of && through_reflection.inverse_of.has_cached_counter?
+              update_counter(-count, through_reflection.inverse_of)
+            else
+              update_counter(-count, through_reflection)
+            end
           else
             update_counter(-count)
           end
```

While this fix works, it seems rather hacky.

The amount of code that deals with counter caches is huge and spread all over the place so it's very hard to follow what's going on. I think this problem would be best tackled by someone that knows the activerecord codebase very well.

**Comment by felixbuenemann** [comment]
To be a little more specific: If you look at the `ActiveRecord::CounterCache` module you will notice that the method `each_counter_cached_associations` only handles belongs_to associations, which means the `_create_record` and `destroy_row` hooks will never work correctly for counter caches on has_many/habtm associations.

**Comment by maclover7** [comment]
@sgrif, thoughts? :grimacing:

**Comment by felixbuenemann** [comment]
I can add a couple more failing test cases if that helps…

**Comment by Tei** [comment]
Can somebody cherry-pick [this commit](https://github.com/rails/rails/commit/025187d9806ddfbdded15d0c7bd8341665ee40e9) from master to 4.2 branch? What do I need to do? Create new issue? PR?

**Comment by piccio** [comment]
I've resolved specifying a custom column name, eg. `:project_ownerships_count` instead of `:projects_count`

**models:**
```
class User < ActiveRecord::Base
  has_many :project_ownerships
  has_many :projects, through: :project_ownerships
end
class Project < ActiveRecord::Base
  has_many :project_ownerships
  has_many :researchers, through: :project_ownerships, source: :user
class ProjectOwnership < ActiveRecord::Base
  belongs_to :project
  belongs_to :user, counter_cache: :project_ownerships_count
end
```
**test:** (with MiniTest)
```
class UserTest < ActiveSupport::TestCase
  it "Counter projects_count should be updated" do
    @researcher.project_ownerships_count.must_equal 0
    @researcher.projects << @project
    @researcher.project_ownerships_count.must_equal 1
    @researcher.projects << @other_project
    @researcher.project_ownerships_count.must_equal 2
  end
end
```

**Comment by matthias-g** [comment]
To me this issue seems still to exist. Using the [gist](https://gist.github.com/ce8a7c9844d579b4f624) mentioned in the first post and replacing line 17 by `gem 'activerecord', '5.2'` the tests should pass, but two of them fail.

This can be fixed by adding the `counter_cache` option to the `has_many` side in Taxon (replacing line 57 by `has_many :classifications, counter_cache: :products_count`).

If this behavior is desired the [guides](http://edgeguides.rubyonrails.org/association_basics.html) should be more clearer than saying:
> You only need to specify the :counter_cache option on the belongs_to side of the association.

Although they also mention that

> This option can be used to configure a custom named :counter_cache. You only need this option when you customized the name of your :counter_cache on the belongs_to association.

**Comment by felixbuenemann** [comment]
@matthias-g You could try the patch from post https://github.com/rails/rails/issues/23739#issuecomment-200465049 above and see if it fixes the issue without having to explicitly specify the counter cache on the has_many side.

**Comment by panasyuk** [comment]
Hi there!
@felixbuenemann  @matthias-g
Have anyone tried to use `dependent: :destroy` option in `has_many`?

```ruby
class Taxon < ActiveRecord::Base
  has_many :classifications
  has_many :products, dependent: :destroy, through: :classifications
end
```

This makes tests green in the [gist](https://gist.github.com/felixbuenemann/ce8a7c9844d579b4f624).
